src/Miunie.Core/Providers/UserReputationProvider.cs

```diff
@@ -17,34 +18,51 @@ public UserReputationProvider(IMiunieUserProvider userProvider, IDateTime dateTi
             _dateTime = dateTime;
         }
 
-        public void AddReputation(MiunieUser invoker, MiunieUser target)
+        public void AddReputation(MiunieUser invoker, MiunieUser target, string reason ="")
         {
             target.Reputation.Value++;
-            target.Reputation.PlusRepLog.TryAdd(invoker.Id, _dateTime.Now);
+            var logEntry = new ReputationLogEntry
+            {
+                Invoker = invoker,
+                Date = _dateTime.Now,
+                Reason = reason,
+                Operation = ReputationOperation.Add
+            };
+            target.Reputation.PlusLog.Add(logEntry);
             _userProvider.StoreUser(target);
         }
 
-        public void RemoveReputation(MiunieUser invoker, MiunieUser target)
+        public void RemoveReputation(MiunieUser invoker, MiunieUser target, string reason = "")
         {
             target.Reputation.Value--;
-            target.Reputation.MinusRepLog.TryAdd(invoker.Id, _dateTime.Now);
+            var logEntry = new ReputationLogEntry
+            {
+                Invoker = invoker,
+                Date = _dateTime.Now,
+                Reason = reason,
+                Operation = ReputationOperation.Remove
+            };
+            target.Reputation.MinusLog.Add(logEntry);
             _userProvider.StoreUser(target);
         }
 
         public bool AddReputationHasTimeout(MiunieUser invoker, MiunieUser target)
-            => HasTimeout(target.Reputation.PlusRepLog, invoker);
+            => HasTimeout(target.Reputation.PlusLog, invoker);
 
         public bool RemoveReputationHasTimeout(MiunieUser invoker, MiunieUser target)
-            => HasTimeout(target.Reputation.MinusRepLog, invoker);
+            => HasTimeout(target.Reputation.MinusLog, invoker);
 
-        private bool HasTimeout(ConcurrentDictionary<ulong, DateTime> log, MiunieUser invoker)
+        private bool HasTimeout(List<ReputationLogEntry> log, MiunieUser invoker)
```

Comment:
ReputationLogEntry is an implementation detail and the Provider shouldn't return it directly.
This is where any phrase resolving might happen and bundling into a simple data object.